Four Rounds RBC Implementation

thehoul opened the pull request:
Implementation according to algorithm 4 in this paper: https://eprint.iacr.org/2021/777.pdf

thehoul committed:
RBC bracha

thehoul committed:
Fix Bracha RBC implementation and add single test

thehoul committed:
Four rounds RBC skeleton

thehoul committed:
renaming 4-Round RBC class

thehoul committed:
Redefine RBC interface and adapt BrachaRBC. Also add a message interface

thehoul committed:
temp: try to implement 4RoundsRBC

thehoul committed:
move bracha to its own folder

thehoul committed:
add custom data structures

thehoul committed:
move and implement RS code

thehoul committed:
update RBC interface

thehoul committed:
export message types to their own file and create protobuf structure

thehoul committed:
move fourround RBC to its own file and implement

add test

thehoul committed:
Temporarily add very basic RS codes and sanity test

thehoul committed:
remove useless structures

thehoul committed:
finish 4RoundsRBC implementation and add unit and a simple "everything goes well" integration test

networking: changed ID to uint32 and add received and sent arrays for testing

reedsolomon: add Encoder and Decoder interfaces, change to use share.Prishare and add simple no error implementation

thehoul committed:
Add 4RoundRBC test with dead nodes

Modify receiver handler to use Receive method instead of AddHandler

rename method in tests, fix non-working and unify timeouts

thehoul committed:
Remove avss files that should not have been here in the first place

thehoul commented:
# Test implemented summary

**Unit**
- [ ] Send a PROPOSE message to a node, checks that it answers correctly
- [ ] Send enough ECHO messages to a node, check that it answers correctly after the correct threshold
- [ ] Send enough READY messages to a node after having sent enough ECHO messages, check that it answers correctly after the correct threshold
- [ ] Send enough READY messages to a node before sending enough ECHO messages, check that it answers correctly after the correct threshold

**Integration**
- [ ] Send a broadcast from a node, and all nodes work fine. Check that all nodes terminate the algorithm, and all agree on the same value. t=2, r=2, nbNodes=7 (3*t+1)
- [ ] Send a broadcast from a node, and one node is dead. Check that all nodes terminate the algorithm, and all agree on the same value. t=2, r=2, nbNodes=7 (3*t+1)
- [ ] Send a broadcast from a node, and two nodes are dead. Check that all nodes terminate the algorithm, and all agree on the same value. t=2, r=2, nbNodes=7 (3*t+1)
- [ ] Send a broadcast from a node, and three nodes are dead. Check that no node terminates the algorithm. t=2, r=2, nbNodes=7 (3*t+1)

thehoul committed:
Add dying dealer test

thehoul committed:
Change logger and allow disabling

AnomalRoil commented:
One test I would add is:
- dealer is connected to a single node (no node id dead, but dealer can only talk to one node), check it still works
- a non-dealer node is connected to a single node (no node is dead again), check it still works

That is: assuming partial connectivity rather than full connectivity.

AnomalRoil commented on rbc/fourRounds/four_rounds_rbc.go:
```diff
@@ -0,0 +1,437 @@
+func (frRbc *FourRoundRBC[M]) handleMessage(instruction *typedefs.Instruction) (error, bool) {
+	var err error = nil
+	finished := false
+	switch op := instruction.Operation.Op.(type) {
```
I'd recommend using the protobuf getters provided by the protobuf implementation below to avoid any risks of nil deref on an invalid/maliciously crafted message

AnomalRoil reviewed (commented):
LGTM so far, but I'm a bit confused with all the Message types you have:
- generic message types
- concrete message types
- protobuf message types

Why can't you rely on just the protobuf types and use them directly?

thehoul committed:
Add stop channel to receive methods to allow halting

Add special error for receiving halted

Add Stop method to RBC

Add tests

Modify MockAuthStream to allow adding delay

thehoul commented on rbc/fourRounds/four_rounds_rbc.go:
```diff
@@ -0,0 +1,437 @@
+		return hashMis
+	})
+
+	// If this value is seen for the first time, add it to T_h and try to reconstruct
```
I am not sure actually. If you look at the algorithm 4 in the [paper](https://eprint.iacr.org/2021/777.pdf), the reconstruction part does not depend on some threshold (unlike other parts that have a `upon...`). It just says from line 16 that when the first ready of some part arrive, it should be stored and line 17 directly attempts to reconstruct. 

I understood this as "try to reconstruct as soon as possible" which is what I am doing here.

thehoul commented on rbc/fourRounds/four_rounds_rbc.go:
```diff
@@ -0,0 +1,437 @@
+	return false
+}
+
+func (frRbc *FourRoundRBC[M]) reconstruct(expHash []byte) ([]M, bool, error) {
+	for ri := 0; ri < frRbc.r; ri++ {
+		if len(frRbc.th) < 2*frRbc.threshold+ri+1 {
```
Yes, of course, if you think this improves readability, I see no reason why not.

thehoul committed:
Change received name for 4Round RBC to f

Remove messaging.go

Add Stop to Bracha RBC and adapt tests

thehoul committed:
fix code analysis warnings

thehoul committed:
Create rbc_test.yml in branch RBC

thehoul committed:
Update rbc_test.yml

thehoul committed:
Update rbc_test.yml

thehoul committed:
rename Bracha test

thehoul committed:
Add locks to prevent concurrency issues with received and sent arrays

thehoul committed:
add log message

Change test sleep timeout

thehoul committed:
Add timeout in RBC test action

thehoul commented:
Left todo

- [ ] Reed Solomon codes
- [ ] Try to use concurrent maps from Go instead of custom ones
- [ ] Multiple instances test
- [ ] Lint

thehoul committed:
remove ConcurrentMaps

thehoul committed:
update RS codes to use Berlekamp-Welch

Modify marshalling and unmarshalling to match new RS code and update test

thehoul commented on rbc/fourRounds/four_rounds_rbc.go:
```diff
@@ -0,0 +1,448 @@
+	f.Lock()
+	defer f.Unlock()
+
+	// Update the count of ready message received for that hash and check if a READY message needs to be sent
```
messages

thehoul commented on rbc/fourRounds/four_rounds_rbc.go:
```diff
@@ -0,0 +1,448 @@
+		err := f.broadcastInstruction(inst)
+		if err != nil {
+			f.log.Err(err).Msg("Failed to broadcast ready message")
+			return false
```
Change method return type to include error, same for other returns below in the method

thehoul commented on rbc/fourRounds/messages.go:
delete

thehoul commented on rbc/fourRounds/messages.proto:
```diff
@@ -0,0 +1,29 @@
+  message Echo{
+    bytes mi=1;
+    bytes h=2;
+    uint32  i=3;
```
maybe int?

thehoul commented on rbc/rbc.go:
```diff
@@ -0,0 +1,47 @@
+	Stop() error
+}
+
+type NodeStoppedError struct{}
```
comment

thehoul commented on reedsolomon/reedsolomoncode.go:
```diff
@@ -0,0 +1,52 @@
+	Decode(msg []infectious.Share) ([]byte, error)
+}
+
+type RSCodes interface {
```
comments

thehoul commented on reedsolomon/reedsolomoncode.go:
```diff
@@ -0,0 +1,52 @@
+	RSDecoder
+}
+
+// BWCodes Berlekamp-Welch implementation of RS codes
+// from this library: https://pkg.go.dev/github.com/Picocrypt/infectious#NewFEC
+type BWCodes struct {
+	RSCodes
+	fec *infectious.FEC
+}
+
+func NewBWCodes(k, n int) *BWCodes {
+	fec, err := infectious.NewFEC(k, n)
+	if err != nil {
+		panic(err)
+	}
+	return &BWCodes{
+		fec: fec,
+	}
+}
+
+func (rs *BWCodes) Encode(msg []byte) ([]infectious.Share, error) {
+	shares := make([]infectious.Share, rs.fec.Total())
+	output := func(s infectious.Share) {
+		shares[s.Number] = s.DeepCopy()
+	}
+
+	err := rs.fec.Encode(msg, output)
+	return shares, err
+}
+
+func (rs *BWCodes) Decode(msg []infectious.Share) ([]byte, error) {
+	res, err := rs.fec.Decode(nil, msg)
+	return res, err
+}
```
Split in its own file

thehoul committed:
removed messages.go

thehoul committed:
remove generic type to 4 rounds RBC

remove dependency to kyber.Scalar in 4Rounds RBC and its test

put RS struct directly as argument

Move Berlekamp-Welch implementation to its own file and modify test accordingly

thehoul committed:
remove generic from RBC interface and remove useless InstanceID type

thehoul committed:
change to pass contexts as discussed in meeting

thehoul committed:
fix fake_network test and data races

fix lint

thehoul committed:
fix lint

change uint32 to int64 to avoid overflow

thehoul committed:
fix warnings and deprecated gomnd

thehoul committed:
fix type check in test

pierluca reviewed (commented):
First chunk of review, working my way through the various files :-) 
More will come...

AnomalRoil commented on reedsolomon/berlekampwelch.go:
```diff
@@ -0,0 +1,53 @@
+package reedsolomon
+
+import (
+	"github.com/HACKERALERT/infectious"
```
Let's avoid pulling that dependency and copy the structs you need instead.

pierluca replied:
Their licensing is non-trivial, I'd be very cautious either ways.
https://github.com/HACKERALERT/infectious/blob/master/LICENSE

thehoul replied:
I'll need help to figure out the licensing, I've never done it before!

AnomalRoil commented on rbc/rbc.go:
```diff
@@ -0,0 +1,50 @@
+func (err NodeStoppedError) Error() string {
+	return "node stopped"
+}
+func (err NodeStoppedError) Is(target error) bool {
+	var nodeStoppedError NodeStoppedError
+	ok := errors.As(target, &nodeStoppedError)
+	return ok
+}
```
Do you really need this, vs just expecting people to use `error.As` or `error.Is` directly??

AnomalRoil commented on rbc/rbc.go:
```diff
@@ -0,0 +1,50 @@
+	Listen(ctx context.Context) error
+}
+
+// NodeStoppedError is used when the Listen or RBroadcast methods are stopped via
+// the Stop method.
+type NodeStoppedError struct{}
+
+func (err NodeStoppedError) Error() string {
+	return "node stopped"
+}
```
I suspect you could just use `errors.New` instead, no?

```suggestion
// NodeStoppedError is used when the Listen or RBroadcast methods are stopped via
// the Stop method.
var NodeStoppedError = errors.New("node stopped")
```

AnomalRoil commented on rbc/fourrounds/messages.proto:
```diff
@@ -0,0 +1,29 @@
+    bytes content=1;
+  }
+
+  message Echo{
+    bytes mi=1;
+    bytes h=2;
+    int64  i=3;
+  }
+
+  message Ready{
+    bytes mi=1;
+    bytes h=2;
+    int64 i=3;
+  }
```
use clearer names

AnomalRoil commented on rbc/bracha/bracha_test.go:
```diff
@@ -0,0 +1,105 @@
+// MockAuthStream mocks an authenticated message stream. Nothing is actually authenticated.
+type MockAuthStream struct {
+	Network    networking.NetworkInterface[[]byte]
+	rcvChan    <-chan []byte
```
Maybe document that you're using unbuffered channels and thus that this is blocking.

AnomalRoil commented on rbc/bracha/bracha.go:
```diff
@@ -0,0 +1,242 @@
+}
+
+func (b *RBC) Listen(ctx context.Context) error {
+	finishedChan := make(chan struct{})
+	b.start(ctx, finishedChan)
+
+	// Wait for the protocol to finish
+	<-finishedChan
```
`b.start` should return the `finishedChan` if that's the way you're supposed to be using it, instead of passing it as a parameter.

AnomalRoil commented on rbc/bracha/bracha.go:
```diff
@@ -0,0 +1,242 @@
+	sync.RWMutex
+}
+
+// NewBrachaRBC creates a new BrachaRBC structure. The marshal un unmarshal methods are used to convert the content
+// of the RBCMessage to and from the value type M
+func NewBrachaRBC(pred func(bool) bool, threshold int, iface rbc.AuthenticatedMessageStream, id uint32) *RBC {
```
>The marshal un unmarshal methods

which ones? looks outdated

AnomalRoil commented on rbc/bracha/bracha_test.go:
```diff
@@ -0,0 +1,105 @@
+	}
+}
+
+func pred(bool) bool {
+	return true
+}
```
Wut? this warrants at least a couple sentences of documentation

AnomalRoil reviewed (changes requested):
Here's some comments

thehoul commented on rbc/bracha/bracha.go:
```diff
@@ -0,0 +1,242 @@
+
+// RBroadcast implements the method from the RBC interface
+func (b *RBC) RBroadcast(ctx context.Context, content bool) error {
+	finishedChan := make(chan struct{})
+	b.start(ctx, finishedChan)
+
+	err := b.startBroadcast(content)
+	if err != nil {
+		return err
+	}
+
+	// Wait for the protocol to finish
+	<-finishedChan
```
That does make a lot more sense. However, I need to know from the caller whether the start function finished because the protocol terminated successfully or because it was stopped (i.e. the context passed was cancelled). In 4Rounds RBC I create two channels, one for success and one for when stopped, and pass both to the start method. Of course, this feels super wrong. Instead, I was thinking of using a channel that returns an error, but then comes the problem of closing the channel; I don't exactly know how to handle this. Also, when listening on a channel, only one of the "listeners" will receive the message (unlike closing, where all listeners will sense that it closed), thus I should document that the channel returned by the start method should only be used by the caller, I suppose.

@AnomalRoil @pierluca  Let me know what you think about this. Also I suppose you haven't reviewed my 4Rounds RBC yet, and that's probably why you did not scream at my `start` implementation there. Feel free to review it first and come back to this comment, it might help understand the context better.

thehoul commented on rbc/bracha/bracha_test.go:
```diff
@@ -0,0 +1,105 @@
+	}
+}
+
+func pred(bool) bool {
+	return true
+}
```
Yup. Reminder for me: add a test to make sure that the implementation reacts correctly when the predicate returns false. Also for 4Rounds RBC

pierluca commented on reedsolomon/berlekampwelch_test.go:
```diff
@@ -0,0 +1,159 @@
+	"go.dedis.ch/kyber/v4/group/edwards25519"
+)
+
+func checkEncoding(t *testing.T, encoded []Encoding, expected []byte, decoder RSDecoder) {
```
as you've seen elsewhere, the `require` (and other libraries) function with a logic of 
`assertion(expected, actual, other parameters)`

Here you might want to do something like
```suggestion
func checkEncoding(t *testing.T, expected []byte, encoded []Encoding, decoder RSDecoder) {
```

pierluca commented on reedsolomon/berlekampwelch_test.go:
```diff
@@ -0,0 +1,159 @@
+	k := 3
+	n := 7
+	rsCodes := NewBWCodes(k, n)
+	// Max error is lower bound of (n-k)/2
```
You might want to move this comment to `Alter some data` and make it explicit that you're altering 2 encoded points, which is the (7-3)/2, and hence that it should still work at that point.

pierluca commented on reedsolomon/berlekampwelch_test.go:
```diff
@@ -0,0 +1,159 @@
+	checkEncoding(t, encoded, message, rsCodes)
+
+	// Add additional corruption and check that it fails (the call to decode fixed the "encoded" array)
+	encoded[1].Val[0] = byte('s')
+	encoded[1].Val[1] = byte('e')
+	encoded[0].Val[0] = byte('x')
+	encoded[0].Val[1] = byte('y')
```
Any reason to repeat this part?

thehoul replied:
yep, the comment just above mentions that the previous call to decode in the `checkEncoding` method fixed the corrupted bytes

pierluca commented on reedsolomon/berlekampwelch_test.go:
```diff
@@ -0,0 +1,159 @@
+	encoded, err := rsCodes.Encode(message)
+	require.NoError(t, err)
+
+	// Drop some data
+	encoded = encoded[2:]
+
+	// Check the encoding
+	checkEncoding(t, encoded, message, rsCodes)
+
+	// Drop 2 more packets
+
+	encoded = encoded[2:]
+
+	checkEncoding(t, encoded, message, rsCodes)
```
Why not just drop all that data at once? Why test twice for some intermediate value that should pass anyways?

pierluca commented on rbc/rbc.go:
```diff
@@ -0,0 +1,41 @@
+
+// NodeStoppedError is used when the Listen or RBroadcast methods are stopped via
+// the Stop method.
+var NodeStoppedError = errors.New("node stopped")
```
I agree with the linter :-) `ErrNodeStopped`

pierluca commented on rbc/rbc.go:
```diff
@@ -0,0 +1,41 @@
+
+// RBC is an interface for an RBC protocol
+type RBC[T any] interface {
+	// RBroadcast blocks until the protocol is finished or an error occurred. The returned bool reflects this results
```
comment is out of data, we're not returning a bool.

pierluca commented on rbc/rbc.go:
```diff
@@ -0,0 +1,41 @@
+}
+
+type AuthenticatedMessageReceiver interface {
+	// Receive blocks until a message is received and returns this message or an error or the given channel
+	// is return to. The channel is used to stop waiting for a message.
+	Receive(context.Context) ([]byte, error)
```
I can't understand the comment, it is also seemingly out of date :-)

pierluca commented on rbc/bracha/bracha.go:
```diff
@@ -0,0 +1,242 @@
+// HandleMsg implements the method from the RBC interface
+func (b *RBC) handleMsg(message Message) (bool, error) {
+	send := false
+	t := PROPOSE
```
what does `t` represent here?

Quoting from Code Complete (a useful bible on software construction):
> Gorla, Benander, and Benander found that the effort required to debug a program was minimized when variables had names that averaged 10 to 16 characters (1990). Programs with names averaging 8 to 20 characters were almost as easy to debug.

pierluca commented on rbc/fourrounds/four_rounds_rbc.go:
```diff
@@ -0,0 +1,408 @@
+	}
+)
+
+type FourRoundRBC struct {
```
This is a huge structure (🚩) with lots of fields having cryptic names (🚩🚩).

May I suggest at least documenting this a little better?

I suspect you might also realize that some fields might be better off being moved into a sub-structure, dealing only with a chunk of the problem.

pierluca commented on rbc/fourrounds/four_rounds_rbc.go:
```diff
@@ -0,0 +1,408 @@
+type FourRoundRBC struct {
+	iface     rbc.AuthenticatedMessageStream
+	predicate func([]byte) bool
+	hash.Hash
```
why does `FourRoundRBC` need to _**embed**_ a `Hash`? 😮

pierluca commented on rbc/fourrounds/four_rounds_rbc.go:
```diff
@@ -0,0 +1,408 @@
+	threshold int
+	sentReady bool
+	sync.RWMutex
+	echoCount   map[string]int
+	readyCounts map[string]int
```
It seems both `echoCount` and `readyCounts` would benefit from being in their own (identical) counter structure, exposing an `Increment(ID)` and `Get(ID)` methods, as opposed to repeating that logic all over the place.

Perhaps `readyMis` could be similarly cleaned up below.

pierluca commented on rbc/fourrounds/four_rounds_rbc.go:
```diff
@@ -0,0 +1,408 @@
+	readyCounts map[string]int
+	readyMis    map[string]map[string]struct{}
+	th          []reedsolomon.Encoding
+	r           int
```
The only case in which it is ever acceptable to have a one-character variable is if you're going to use it on the next 1-2 lines and never again. This is not the case here, you might want to give it a proper name :-)

pierluca commented on rbc/fourrounds/four_rounds_rbc.go:
```diff
@@ -0,0 +1,408 @@
+	readyMis    map[string]map[string]struct{}
+	th          []reedsolomon.Encoding
+	r           int
+	finalValue  []byte
+	finished    bool
```
It seems `finished` is strictly equivalent to `finalValue != nil`.
If I understand this correctly, you might want to remove the redundant boolean field.
If I'm mistaken, you might want to use a separate structure which only allows setting both atomically, to avoid a case where `finalValue` is not set but `finished` is.

pierluca commented on rbc/fourrounds/four_rounds_rbc.go:
Since some of the logic is the same as the `Bracha` RBC, would it make sense to extend its logic and override the necessary parts? 🤔 

Also, filename-wise, `fourrounds` is already in your package name, you might just want to call this `rbc.go`

pierluca commented on rbc/fourrounds/four_rounds_rbc.go:
```diff
@@ -0,0 +1,408 @@
+	}
+}
+
+func (f *FourRoundRBC) FreshHash(bs []byte) ([]byte, error) {
+	f.Hash.Reset()
+
+	// Write the bytes
+	_, err := f.Hash.Write(bs)
+	if err != nil {
+		return nil, err
+	}
+	h := f.Hash.Sum(nil)
+	return h, nil
+}
```
You've marked this method as public. Is this never accessed concurrently?
That expectation is not documented. Otherwise you could have problems here.

pierluca commented on rbc/fourrounds/four_rounds_rbc.go:
```diff
@@ -0,0 +1,408 @@
+func (f *FourRoundRBC) start(ctx context.Context) chan error {
+	returnChan := make(chan error)
+	go func() {
+		for {
```
why not loop while `finished` is not true, and have a single exit point from this function?

pierluca commented on rbc/fourrounds/four_rounds_rbc.go:
```diff
@@ -0,0 +1,408 @@
+				continue
+			}
+			if finished {
+				f.log.Err(err).Msg("Protocol terminated")
+				returnChan <- err
```
I'm not sure you want to log an error when you know `err == nil` here.
Similarly, you might just want to do `close(returnChan)` as opposed to passing a `nil` in.
Finally, as mentioned above, you might want to do that at the end of the `go func()` for clarity.

pierluca commented on rbc/fourrounds/four_rounds_rbc.go:
```diff
@@ -0,0 +1,408 @@
+	}
+
+	err = <-finishedChan
+	close(finishedChan) // TODO this should not be closed by the callee (?)
```
Correct :-) you can remove this.

pierluca commented on rbc/fourrounds/four_rounds_rbc.go:
```diff
@@ -0,0 +1,408 @@
+	finishedChan := f.start(ctx)
+
+	err := <-finishedChan
+	close(finishedChan) // TODO this should not be closed by the callee (?)
```
Correct :-) you can remove this.

pierluca commented on rbc/fourrounds/four_rounds_rbc.go:
```diff
@@ -0,0 +1,408 @@
+
+func (f *FourRoundRBC) receivePropose(msg *typedefs.Message_Propose) error {
+	if !f.predicate(msg.GetContent()) {
+		return xerrors.New("Given value did not pass the predicate")
```
you might want to define a global `ErrPredicateRejected` or similar, making things easier to test.

pierluca commented on rbc/fourrounds/four_rounds_rbc.go:
```diff
@@ -0,0 +1,408 @@
+		f.log.Printf("Sent Ready message for %x", msg.GetMi())
+	}
+
+	var value []byte
+	hashMis, ok := f.readyMis[string(msg.GetH())]
+
+	if !ok {
+		hashMis = make(map[string]struct{})
+	}
+	_, notFirst := hashMis[string(msg.GetMi())]
+	firstTime := !notFirst
+
+	// Put the value in the map
+	hashMis[string(msg.GetMi())] = struct{}{}
+	// Update the map
+	f.readyMis[string(msg.GetH())] = hashMis
```
conceptually, all of this is an implementation detail that should be abstracted away behind an API that provides a single call `f.readyMis.IsSeenForFirstTime(msg.GetH())` (assuming `readyMis` is the best name for this, but I assume there might be another one)

pierluca commented on rbc/fourrounds/four_rounds_rbc.go:
```diff
@@ -0,0 +1,408 @@
+		}
+	}
+
+	if finished {
+		f.finalValue = value
+		f.finished = true
+		return true, nil
+	}
```
This can only ever be true within `if firstTime {}`, why keep it outside of that `if` ?
You can then remove the variable definition at lines 294/308/309 and define+assign them at line 320 instead.

pierluca commented on rbc/fourrounds/four_rounds_rbc.go:
```diff
@@ -0,0 +1,408 @@
+	return f.iface.Broadcast(out)
+}
+
+func createReadyMessage(mi, h []byte, i int64) *typedefs.Instruction {
+	readyMsg := &typedefs.Message_Ready{
+		Mi: mi,
+		H:  h,
+		I:  i,
+	}
+	msg := &typedefs.Message{Op: &typedefs.Message_ReadyInst{ReadyInst: readyMsg}}
+	inst := &typedefs.Instruction{Operation: msg}
+	return inst
+}
+
+func createEchoMessage(mi, h []byte, i int64) *typedefs.Instruction {
+	echoMsg := &typedefs.Message_Echo{
+		Mi: mi,
+		H:  h,
+		I:  i,
+	}
+	msg := &typedefs.Message{Op: &typedefs.Message_EchoInst{EchoInst: echoMsg}}
+	inst := &typedefs.Instruction{Operation: msg}
+	return inst
+}
+
+func createProposeMessage(ms []byte) *typedefs.Instruction {
+	proposeMsg := &typedefs.Message_Propose{
+		Content: ms,
+	}
+	msg := &typedefs.Message{Op: &typedefs.Message_ProposeInst{ProposeInst: proposeMsg}}
+	inst := &typedefs.Instruction{Operation: msg}
+	return inst
+}
```
This is very nice because you abstract away the protobuf stuff from the logic itself.
You should adopt the same logic elsewhere (when receiving data), as you might not want your code to be entirely tied to protobuf.

pierluca commented on rbc/fourrounds/four_rounds_rbc_test.go:
I haven't gotten around to reviewing this file yet.

thehoul commented on rbc/bracha/bracha.go:
```diff
@@ -0,0 +1,242 @@
+	}
+
+	// Wait for the protocol to finish
+	err = <-finishedChan
+	close(finishedChan)
+	return err
```
Great! I thought I had to close them. I happy just to remove this. I added the close there in order not to forget to do it and added ToDos to remind me.

thehoul committed:
fix PR reviews

thehoul committed:
fix lint

thehoul committed:
use clearer names in message.proto for 4Rounds RBC

thehoul committed:
rename bracha and 4rounds to rbc because their package name already reflect this

thehoul committed:
adapt 4rounds RBC test to the new networking

thehoul committed:
add test on UDP with small network

thehoul committed:
fix lint

thehoul committed:
fix lint

thehoul committed:
get logger from logging package

thehoul committed:
reduce logging

thehoul committed:
change reed solomon interface to use pointers

thehoul committed:
send one broadcast with all shares instead of one broadcast for each

improve tests and add UDP network small and stress tests

thehoul committed:
add benchmark test to get data

thehoul committed:
fix lint

thehoul committed:
fix RBC receive READY tests

improve RBC to use less maps

refactor a few names and add comments

thehoul committed:
remove weird mutex in networking

thehoul committed:
update benchmark csv

thehoul committed:
fix lint

add skip for benchmark test to avoid re-benching every time

thehoul committed:
remove unused r in 4Rounds RBC

thehoul committed:
add benchmark

thehoul committed:
fix lint

thehoul committed:
multi RBC - stress tests don't pass

thehoul committed:
fix lint

thehoul committed:
Better channel handling

thehoul committed:
of course I forgot to fix lint

thehoul committed:
add multiple instances test

thehoul committed:
update Bracha implementation to new interface

thehoul committed:
Update RBC interface

thehoul committed:
add Instance interface to RBC

thehoul committed:
add padding for message of size different from multiple of threshold+1

thehoul committed:
make RBC broadcast return the instance created

thehoul committed:
add node struct

add general protobuf Packet message type for all ADKG messages

thehoul committed:
add node struct

add general protobuf Packet message type for all ADKG messages

thehoul committed:
add node struct

add general protobuf Packet message type for all ADKG messages

thehoul committed:
add GetInstances and SetPredicate to the RBC interface

thehoul committed:
udpate test benchmark to add message bytes count

thehoul committed:
fix lint

thehoul committed:
fix tests and skip benchmark

thehoul committed:
simplify test
